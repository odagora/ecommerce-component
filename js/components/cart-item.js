import { formattedPrice } from "../utils/formattedPrice.js"

export function cartItem({image, altText, name, price, subTotal, quantity}) {
  const newPrice = formattedPrice(price * 100)
  const newSubTotal = formattedPrice(subTotal)

  return `
    <li>
      <div class="plate">
        <img src="${image}" alt="${altText}" class="plate" />
        <div class="quantity">${quantity}</div>
      </div>
      <div class="content">
        <p class="menu-item">${name}</p>
        <p class="price">${newPrice}</p>
      </div>
      <div class="quantity__wrapper">
        <button class="decrease">
          <img src="images/chevron.svg" />
        </button>
        <div class="quantity">${quantity}</div>
        <button class="increase">
          <img src="images/chevron.svg" />
        </button>
      </div>
      <div class="subtotal">${newSubTotal}</div>
      <button class="remove">Remove</button>
    </li>
  `
}