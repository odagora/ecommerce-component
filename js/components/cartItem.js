// Importing necessary event constants and utility functions
import { CartEvents } from "../events/cartEvents.js";
import { numberToPrice } from "../utils/format.js";

// Define a custom element named 'cart-item' extending from the HTMLElement class
export class CartItem extends HTMLElement {
  #state;
  // Constructor takes an object with properties like name, image, alt, quantity, price, and subTotal
  constructor({ name, image, alt, quantity, price, subTotal }) {
    super();
    // Initialize properties with the values passed in the constructor
    this.name = name;
    this.image = image;
    this.alt = alt;
    this.quantity = quantity;
    this.price = price;
    this.isRemoveCTAHidden = true;
    // Set subTotal to the provided value or default to the product price
    this.subTotal = subTotal ?? this.price;
    this.#state = {
      quantity: this.quantity,
      subTotal: this.subTotal
    }
  }

  // Getter method to get the value of a property in the state object
  getState(path) {
    return this.#state[path];
  }

  // Setter method to set the value of a property in the state object
  setState(path, value) {
    if (this.#state[path] !== value) {
      this.#state = { ...this.#state, [path]: value };
    }
  }

  // ConnectedCallback is called when the element is inserted into the DOM
  connectedCallback() {
    // Render the initial state
    this.render();
    // Add event listeners for remove, decrease, and increase buttons
    this.addEventListener('click', this.#handleButtonClick)
  }

  // DisconnectedCallback is called when the element is removed from the DOM
  disconnectedCallback() {
    // Remove previously added event listeners to prevent memory leaks
    this.removeEventListener('click', this.#handleButtonClick)
  }

  // Private method to handle click events on the remove, decrease, and increase buttons
  #handleButtonClick(event) {
    const { target } = event;
    const targetClassList = target.closest('button').classList;

    if (targetClassList.contains('remove')) {
      this.#removeFromCartEvent();
    } else if (targetClassList.contains('decrease')) {
      this.#decreaseQuantity();
      this.#quantityChangeEvent('decrease-quantity');
    } else if (targetClassList.contains('increase')) {
      this.#increaseQuantity();
      this.#quantityChangeEvent('increase-quantity');
    }
  }

  // Private method to increase the quantity and update the subTotal property in the state object.
  #increaseQuantity() {
    const currentValue = this.getState('quantity');
    const newValue = currentValue + 1;
    this.setState('quantity', newValue);
    this.setState('subTotal', newValue * this.price);
    this.isRemoveCTAHidden = true;
    this.render();
  }

  // Private method to decrease the quantity and update the subTotal property in the state object.
  #decreaseQuantity() {
    const currentValue = this.getState('quantity');
    const newValue = currentValue - 1;
    if (currentValue > 0) {
      this.setState('quantity', newValue);
      this.setState('subTotal', newValue * this.price);
    }
    if (this.getState('quantity') === 0) {
      this.isRemoveCTAHidden = false;
    }
    this.render();
  }

  // Private method to dispatch a custom event when 'Remove from cart' button is clicked
  #removeFromCartEvent() {
    document.dispatchEvent(CartEvents.removeFromCart(this.name));
  }

  // Private method to dispatch custom events for 'Decrease quantity' and 'Increase quantity'
  #quantityChangeEvent(event) {
    const productDetails = {
      name: this.name,
      quantity: this.quantity,
      price: this.price
    };
    if (event === 'decrease-quantity') {
      document.dispatchEvent(CartEvents.decreaseQuantity(productDetails));
    }

    if (event === 'increase-quantity') {
      document.dispatchEvent(CartEvents.increaseQuantity(productDetails));
    }
  }

  // Render method to create the initial HTML structure of the component
  render() {
    this.innerHTML = `
      <li data-name="${this.name}">
        <div class="plate">
          <img src="images/${this.image}" alt="${this.alt}" class="plate" />
          <div class="quantity">${this.getState('quantity')}</div>
        </div>
        <div class="content">
          <p class="menu-item">${this.name}</p>
          <p class="price">${numberToPrice(this.price)}</p>
        </div>
        <div class="quantity__wrapper">
          <button class="decrease">
            <img src="images/chevron.svg" />
          </button>
          <div class="quantity">${this.getState('quantity')}</div>
          <button class="increase">
            <img src="images/chevron.svg" />
          </button>
        </div>
        <div class="subtotal">${numberToPrice(this.getState('subTotal'))}</div>
        <button class="remove ${this.isRemoveCTAHidden ? 'hidden' : ''}">Remove</button>
      </li>
    `;
  }
}

// Define the custom element using window.customElements
if (!customElements.get('cart-item')) {
  customElements.define('cart-item', CartItem);
}
