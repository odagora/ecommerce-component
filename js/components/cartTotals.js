import { numberToPrice } from "../utils/format.js";

// Defining a custom element called cart-totals
export class CartTotals extends HTMLElement {
  // Constructor takes and object with subTotal, tax, and total properties
  constructor({subTotal, tax, total}) {
    super();
    this.subTotal = subTotal ?? 0;
    this.tax = tax ?? 0;
    this.total = total ?? 0;
  }

  // ConnectedCallback is called when the element is inserted into the DOM
  connectedCallback() {
    // Initial rendering
    this.render();
  }

  // Function to update cart totals
  update() {
    const subtotal = this.querySelector('.subtotal');
    const tax = this.querySelector('.tax');
    const total = this.querySelector('.total .price');

    subtotal.innerText = numberToPrice(this.subTotal);
    tax.innerText = numberToPrice(this.tax);
    total.innerText = numberToPrice(this.total);
  }

  // Render method to create the initial HTML structure of the component
  render() {
    this.innerHTML = `
    <div class="line-item">
      <div class="label">Subtotal:</div>
      <div class="amount price subtotal" data-testid="subtotal">${numberToPrice(this.subTotal)}</div>
    </div>
    <div class="line-item">
      <div class="label">Tax:</div>
      <div class="amount price tax" data-testid="tax">${numberToPrice(this.tax)}</div>
    </div>
    <div class="line-item total">
      <div class="label">Total:</div>
      <div class="amount price total" data-testid="total">${numberToPrice(this.total)}</div>
    </div>
    `
  }
}

if (!customElements.get('cart-totals')) {
  customElements.define('cart-totals', CartTotals);
}

