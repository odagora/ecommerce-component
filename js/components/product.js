import { formattedPrice } from "../utils/formattedPrice.js"

export function product({image, alt, name, price}) {
  const newPrice = formattedPrice(price)

  return `
    <li>
      <div class="plate">
        <img src="${`images/${image}`}" alt="${alt}" class="plate" />
      </div>
      <div class="content">
        <p class="menu-item">${name}</p>
        <p class="price">${newPrice}</p>
        <button class="add">Add to cart</button>
      </div>
    </li>
  `
}