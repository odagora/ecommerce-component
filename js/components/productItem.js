import { ProductEvents } from "../events/productEvents.js";
import { numberToPrice } from "../utils/format.js"

// Define a custom element named 'product-item' extending from the HTMLElement class
export class ProductItem extends HTMLElement {
  // Constructor takes an object with properties like image, alt, name, and price
  constructor({ image, alt, name, price }) {
    super();
    // Initialize properties with the values passed in the constructor
    this.image = image;
    this.alt = alt;
    this.name = name;
    this.price = price;
  }

  // ConnectedCallback is called when the element is inserted into the DOM
  connectedCallback() {
    // Render the initial state
    this.render();
    // Add an event listener to the 'Add to cart' button
    this.querySelector('button').addEventListener('click', () => {
      // Call a private method to handle the 'Add to cart' event
      this.#addToCartEvent();
    });
  }

  // DisconnectedCallback is called when the element is removed from the DOM
  disconnectedCallback() {
    // Remove the previously added event listener to prevent memory leaks
    this.querySelector('button').removeEventListener('click', () => {
      this.#addToCartEvent();
    });
  }

  // Private method to dispatch a custom event when 'Add to cart' button is clicked
  #addToCartEvent() {
    document.dispatchEvent(
      ProductEvents.addToCart({
        name: this.name,
        price: this.price,
        image: this.image,
        alt: this.alt,
        quantity: 1,
      })
    );
  }

  // Update method to toggle button class and update button text/content
  update() {
    const button = this.querySelector('button');
    button.classList.toggle('add');
    button.classList.toggle('in-cart');

    if (button.classList.contains('in-cart')) {
      // If the item is in the cart, show a checkmark image
      const image = document.createElement('img');
      image.src = 'images/check.svg';
      image.alt = 'check';
      button.innerText = 'In Cart';
      button.appendChild(image);
      return;
    }

    // If the item is not in the cart, revert to default button text
    button.innerText = 'Add to cart';
  }

  // Render method to create the initial HTML structure of the component
  render() {
    this.innerHTML = `
      <li data-name="${this.name}">
        <div class="plate">
          <img src="${`images/${this.image}`}" alt="${this.alt}" class="plate" />
        </div>
        <div class="content">
          <p class="menu-item">${this.name}</p>
          <p class="price">${numberToPrice(this.price)}</p>
          <button class="add">Add to cart</button>
        </div>
      </li>
    `;
  }
}

// Define the custom element using window.customElements
window.customElements.define('product-item', ProductItem);
