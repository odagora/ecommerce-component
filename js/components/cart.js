// Importing necessary classes for cart items and totals
import { CartItem } from "./cartItem.js";
import { CartTotals } from "./cartTotals.js";

// Define a class named 'Cart'
export class Cart {
  // Constructor initializes properties for items, subTotal, tax, total, and totalElement
  constructor() {
    this.items = [];
    this.subTotal = 0;
    this.tax = 0;
    this.total = 0;
    // Create an instance of CartTotals with initial values and assign it to totalElement
    this.totalElement = new CartTotals(this.subTotal, this.tax, this.total);
  }

  // Method to add a product to the cart
  addProduct(productDetails) {
    // Create a new CartItem instance with the provided product details
    const cartItem = new CartItem(productDetails);
    // Add the new cart item to the items array
    this.items.push(cartItem);
    // Render the updated cart
    this.render();
  }

  // Method to remove a product from the cart
  removeProduct(productDetails) {
    // Filter out the item with the specified name from the items array
    const newItems = this.items.filter(item => item.name != productDetails.name);
    this.items = newItems;
    // Render the updated cart
    this.render();
  }

  // Method to update the quantity and subTotal of a product in the cart
  updateProduct(productDetails, quantity, subTotal) {
    // Find the index of the product in the items array
    const productIndex = this.items.findIndex(item => item.name == productDetails.name);
    // Update quantity and subTotal for the specified product
    this.items[productIndex].quantity = quantity;
    this.items[productIndex].subTotal = subTotal;
    // Render the updated cart
    this.render();

    // If quantity is 0, update the call-to-action (CTA) in the cart item
    if (quantity === 0) {
      this.items[productIndex].updateCTA();
    }
  }

  // Method to render the totals section of the cart
  renderTotals() {
    const totalContainer = document.querySelector('.totals');
    // Append the totalElement to the totals container
    totalContainer.appendChild(this.totalElement);
  }

  // Method to update the totals section with new values
  updateTotals(subTotal, tax, total) {
    // Update the properties of totalElement with new values
    this.totalElement.subTotal = subTotal;
    this.totalElement.tax = tax;
    this.totalElement.total = total;
    // Call the update method of totalElement to reflect the changes in the UI
    this.totalElement.update();
  }

  // Method to check if the cart is empty
  isEmpty() {
    const cartItems = document.querySelectorAll('.cart-summary li');
    return !cartItems.length;
  }

  // Method to render the entire cart, including items and totals
  render() {
    const cartContainer = document.querySelector('.cart-summary');
    const totalContainer = document.querySelector('.totals');
    // Clear the existing content in the cart and totals containers
    cartContainer.innerHTML = '';
    totalContainer.innerHTML = '';
    // Append each cart item to the cart container
    this.items.forEach(item => {
      cartContainer.appendChild(item);
    });
    // Render the totals section
    this.renderTotals();
  }
}
