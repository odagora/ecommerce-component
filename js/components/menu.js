import { ProductItem } from "./productItem.js";

export class Menu {
  constructor(products) {
    this.products = products;
  }

  render() {
    const menuContainer = document.querySelector('.menu');
    menuContainer.innerHTML = '';

    this.products.forEach((product) => {
      const productItem = new ProductItem({ ...product });
      menuContainer.appendChild(productItem);
    })
  }
}