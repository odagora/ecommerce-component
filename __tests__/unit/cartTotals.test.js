import { CartTotals } from "../../js/components/cartTotals";

describe("CartTotals", () => {
  let cartTotals, subTotal, tax, total;

  beforeEach(() => {
    // Arrange
    subTotal = 10;
    tax = 1;
    total = 11;

    // Act
    cartTotals = new CartTotals({ subTotal, tax, total });
    document.body.appendChild(cartTotals);
  });

  test("should create \'CartTotals\' element with correct properties", () => {
    // Assert
    expect(cartTotals.subTotal).toBe(subTotal);
    expect(cartTotals.tax).toBe(tax);
    expect(cartTotals.total).toBe(total);
  });

  test('should render the correct HTML', () => {
    // Assert
    // Use of snapshot testing to avoid specifying the expected UI structure. See 'https://jestjs.io/docs/snapshot-testing'
    expect(cartTotals.innerHTML).toMatchSnapshot();
  })

  test('should update the cart totals when the update function is called', () => {
    // Arrange
    const newSubTotal = 20;
    const newTax = 2;
    const newTotal = 22;

    // Act
    cartTotals.subTotal = newSubTotal;
    cartTotals.tax = newTax;
    cartTotals.total = newTotal;
    cartTotals.update();

    // Assert
    expect(cartTotals.subTotal).toBe(newSubTotal);
    expect(cartTotals.tax).toBe(newTax);
    expect(cartTotals.total).toBe(newTotal);
  })
});