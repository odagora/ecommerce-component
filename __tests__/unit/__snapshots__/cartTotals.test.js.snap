// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartTotals should render the correct HTML 1`] = `
"
    <div class="line-item">
      <div class="label">Subtotal:</div>
      <div class="amount price subtotal">$0.10</div>
    </div>
    <div class="line-item">
      <div class="label">Tax:</div>
      <div class="amount price tax">$0.01</div>
    </div>
    <div class="line-item total">
      <div class="label">Total:</div>
      <div class="amount price total">$0.11</div>
    </div>
    "
`;
